### PR TITLE
Migrate llvm::Optional to std::optional

### PR DESCRIPTION
**Pull request by bnbarham**
LLVM has removed llvm::Optional, move over to std::optional. Also clang-format to fix up all the renamed #includes.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by compnerd — APPROVED** [review]
Seems pretty mechanical.  Thank you for `clang-formatting` the patch!

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by al45tair** [comment]
(On Windows) looks like maybe a missing `#include <optional>` somewhere:

```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\include\swift/Basic/Lazy.h(36,3): error: no template named 'optional' in namespace 'std'; did you mean 'llvm::Optional'?
   36 |   std::optional<T> Value;
      |   ^~~~~~~~~~~~~
      |   llvm::Optional
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\stdlib\include\llvm/ADT/Optional.h(251,29): note: 'llvm::Optional' declared here
  251 | template <typename T> class Optional {
      |                             ^
1 error generated.
[107/308][ 34%][6.689s] Building CXX object stdlib\public\CMakeFiles\swiftDemangling-windows-x86_64.dir\__\__\lib\Demangling\Demangler.cpp.obj
[108/308][ 35%][6.792s] Building CXX object stdlib\public\runtime\CMakeFiles\swiftRuntime-windows-x86_64.dir\GenericMetadataBuilder.cpp.obj
In file included from C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\GenericMetadataBuilder.cpp:17:
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\include\swift/Runtime/GenericMetadataBuilder.h(400,11): warning: unused variable 'pattern' [-Wunused-variable]
  400 |     auto *pattern = patternBuffer.ptr;
      |           ^~~~~~~
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\stdlib\public\runtime\GenericMetadataBuilder.cpp(296,25): note: in instantiation of member function 'swift::GenericMetadataBuilder<InProcessReaderWriter>::buildGenericValueMetadata' requested here
  296 |   auto result = builder.buildGenericValueMetadata({description}, arguments,
      |                         ^
1 warning generated.
```

**Review by al45tair — APPROVED** [review]
Looks pretty mechanical. I've skimmed over most of it, but what I looked at all seemed good. There's a Windows issue, but that looks like a missing `<optional>` somewhere (probably some difference in the C++ library that causes it to be pulled in on the other platforms but not on the version we're using on Windows).

I'll be especially pleased to see the back of `llvm::None` as *that* causes problems adding `None` cases to enums.

**Review by mikeash — APPROVED** [review]
That's a big ol' diff. I didn't read it all, but assuming it's all `s/llvm::Optional/std::optional/` then I'm happy.

**Comment by bnbarham** [comment]
> Looks pretty mechanical. I've skimmed over most of it, but what I looked at all seemed good. There's a Windows issue, but that looks like a missing <optional> somewhere (probably some difference in the C++ library that causes it to be pulled in on the other platforms but not on the version we're using on Windows).

Yeah, MSVC doesn't end up pulling in `<optional>` from `<functional>` like macOS does (I didn't check where it's pulled in on Linux). I can easily fix this by adding `<optional>` into `Basic/Lazy.h` but I find it strange that the stdlib uses `Basic` at all TBH - I certainly wouldn't expect a `Basic` change to possibly impact ABI (and no, I don't know if we *are* using it in ABI somewhere, it just makes it very easy to).

> I'll be especially pleased to see the back of llvm::None as that causes problems adding None cases to enums.

Heh. So just to be clear - this still doesn't remove it from a bunch of directories, namely `ABI/Demangling/Remote*/Threading`. These end up being used from the stdlib (and maybe ABI somewhere) as well as from tools. Technically those are all currently using `std::optional` though since that's what `llvm::Optional` is aliased to on `main`.

> That's a big ol' diff. I didn't read it all, but assuming it's all s/llvm::Optional/std::optional/ then I'm happy.

If only. Unfortunately:
1. More `Optional` were added with a `using namespace llvm`
2. SourceKit and unittests both had the unqualified versions still
3. There's also `None` and `NoneType`

But... close enough :)

**Review by xedin — APPROVED** [review]
Went through most of the Sema.

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -813,17 +813,17 @@ class FunctionArgApplyInfo {
...
-  static llvm::Optional<FunctionArgApplyInfo>
+  static std::optional<FunctionArgApplyInfo>
   get(ArgumentList *argList, Expr *argExpr, unsigned argIdx, Type argType,
       unsigned paramIdx, Type fnInterfaceType, FunctionType *fnType,
       const ValueDecl *callee) {
     assert(fnType);
 
     if (argIdx >= argList->size())
-      return llvm::None;
+    return std::nullopt;
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply include/swift/Sema/ConstraintSystem.h:823]: That is very concerning. To be clear, I didn't edit any of these files manually. This was a simple `sed` + `git clang-format`. I'll look into it. Thanks for taking such a close look!

[the remark this answers is not in the document]
Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h:823]: No worries!

[the remark this answers is not in the document]
Comment (bnbarham) [reply include/swift/Sema/ConstraintSystem.h]: I updated clang-format and that appears to have fixed these

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Review by artemcm — APPROVED** [review]

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
https://github.com/apple/llvm-project/pull/8120

@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please smoke test

**Comment by bnbarham** [comment]
@swift-ci please smoke test

**Comment by bnbarham** [comment]
@swift-ci please test Windows platform

**Pull request merged**